src/chainstate/stacks/index/storage.rs

```diff
@@ -0,0 +1,1213 @@
+}
+
+pub fn fseek<F: Seek>(f: &mut F, off: u64) -> Result<u64, Error> {
+    f.seek(SeekFrom::Start(off))
+        .map_err(Error::IOError)
+}
+
+pub fn fseek_end<F: Seek>(f: &mut F) -> Result<u64, Error> {
+    f.seek(SeekFrom::End(0))
+        .map_err(Error::IOError)
+}
+
+pub fn read_all<R: Read>(f: &mut R, buf: &mut [u8]) -> Result<usize, Error> {
+    let mut cnt = 0;
+    while cnt < buf.len() {
+        let nr = f.read(&mut buf[cnt..])
+            .map_err(Error::IOError)?;
+
+        if nr == 0 {
+            break;
+        }
+
+        cnt += nr;
+    }
+    Ok(cnt)
+}
+
+pub fn write_all<W: Write>(f: &mut W, buf: &[u8]) -> Result<usize, Error> {
+    let mut cnt = 0;
+    while cnt < buf.len() {
+        let nw = f.write(&buf[cnt..buf.len()])
+            .map_err(Error::IOError)?;
+        cnt += nw;
+    }
+    Ok(cnt)
+}
+
+/// In-RAM trie storage.
+/// Used by TrieFileStorage to buffer the next trie being built.
+pub struct TrieRAM {
+    data: Vec<(TrieNodeType, TrieHash)>,
+    block_header: BlockHeaderHash,
+    readonly: bool,
+
+    read_count: u64,
+    read_backptr_count: u64,
+    read_node_count: u64,
+    read_leaf_count: u64,
+
+    write_count: u64,
+    write_node_count: u64,
+    write_leaf_count: u64,
+
+    total_bytes: usize
+}
+
+// Trie in RAM without the serialization overhead
+impl TrieRAM {
+    pub fn new(block_header: &BlockHeaderHash, capacity_hint: usize) -> TrieRAM {
+        TrieRAM {
+            data: Vec::with_capacity(capacity_hint),
+            block_header: block_header.clone(),
+            readonly: false,
+
+            read_count: 0,
+            read_backptr_count: 0,
+            read_node_count: 0,
+            read_leaf_count: 0,
+
+            write_count: 0,
+            write_node_count: 0,
+            write_leaf_count: 0,
+
+            total_bytes: 0
+        }
+    }
+
+    #[cfg(test)]
+    #[allow(dead_code)]
+    pub fn stats(&mut self) -> (u64, u64) {
+        let r = self.read_count;
+        let w = self.write_count;
+        self.read_count = 0;
+        self.write_count = 0;
+        (r, w)
+    }
+
+    #[cfg(test)]
+    #[allow(dead_code)]
+    pub fn node_stats(&mut self) -> (u64, u64, u64) {
+        let nr = self.read_node_count;
+        let br = self.read_backptr_count;
+        let nw = self.write_node_count;
+
+        self.read_node_count = 0;
+        self.read_backptr_count = 0;
+        self.write_node_count = 0;
+         
+        (nr, br, nw)
+    }
+
+    #[cfg(test)]
+    #[allow(dead_code)]
+    pub fn leaf_stats(&mut self) -> (u64, u64) {
+        let lr = self.read_leaf_count;
+        let lw = self.write_leaf_count;
+
+        self.read_leaf_count = 0;
+        self.write_leaf_count = 0;
+
+        (lr, lw)
+    }
+
+    /// Walk through the bufferred TrieNodes and dump them to f.
+    fn dump_traverse<F: Read + Write + Seek>(&mut self, f: &mut F, root: &TrieNodeType, hash: &TrieHash, parent_hash: &BlockHeaderHash) -> Result<u64, Error> {
```

Comment:
This function has to do a lot of cloning, since it's non-destructive, but `.dump` is only ever called from a destructive context (that is, after being flushed, the object is thrown away anyways). According to `perf`, almost as much time is spent in memmove as in computing hashes, which leads me to believe it may be worthwhile to try and cut down on some clones. What do you think? We can also put this off to a separate performance improvement PR.

Reply:
I didn't spend any time optimizing this path because the runtime cost ought to be dominated by the cost of writing the data to disk and `fsync`'ing it.  Also, a miner can call `.dump()` concurrently with searching for the next block, since it only needs to calculate MARF root hash to finish validating the block (per SIP 005).

Regardless, making this faster will make chain downloads go faster, so it's worthwhile to do at some point.  However, I'll defer it until we spend some time doing end-to-end optimization of the chain synchronization process.